docs/specifications/url.rst

```diff
@@ -0,0 +1,112 @@
+
+* the base32-encoded SHA1 hash of the DER form of an x509v3 certificate
+* zero or more network addresses
+* an object identifier
+
+A Foolscap client tries to connect to each network address in turn.
+If a connection is established then TLS is negotiated.
+The server is authenticated by matching its certificate against the hash in the fURL.
+A matching certificate serves as proof that the handshaking peer is the correct server.
+This serves as the process by which the client authenticates the server.
+
+The client can then exercise further Foolscap functionality using the fURL's object identifier.
+If the object identifier is an unguessable, secret string then it serves as a capability.
+This serves as the process by which the server authorizes the client.
+
+NURLs
+-----
+
+The authentication and authorization properties of fURLs are a good fit for Tahoe-LAFS' requirements.
+These are not inherently tied to the Foolscap protocol itself.
+In particular they are beneficial to :doc:`../proposed/http-storage-node-protocol` which uses HTTP instead of Foolscap.
+It is conceivable they will also be used with WebSockets at some point as well.
+
+Continuing to refer to these URLs as fURLs when they are being used for other protocols may cause confusion.
+Therefore,
+this document coins the name **NURL** for these URLs.
+This can be considered to expand to "**N**\ ew URLs" or "Authe\ **N**\ ticating URLs" or "Authorizi\ **N**\ g URLs" as the reader prefers.
+
+Syntax
+------
+
+The EBNF for a NURL is as follows::
+
+  nurl         = scheme, hash, "@", net-loc-list, "/", swiss-number, [ version1 ]
+
+  scheme       = "pb://"
+
+  hash         = unreserved
+
+  net-loc-list = net-loc, [ { ",", net-loc } ]
+  net-loc      = hostname, [ ":" port ]
+  hostname     = domain | IPv4address | IPv6address
+
+  swiss-number = segment
+
+  version1     = "#v=1"
+
+See https://tools.ietf.org/html/rfc3986#section-3.3 for the definition of ``segment``.
+See https://tools.ietf.org/html/rfc2396#appendix-A for the definition of ``unreserved``.
+See https://tools.ietf.org/html/draft-main-ipaddr-text-rep-02#section-3.1 for the definition of ``IPv4address``.
+See https://tools.ietf.org/html/draft-main-ipaddr-text-rep-02#section-3.2 for the definition of ``IPv6address``.
+See https://tools.ietf.org/html/rfc1035#section-2.3.1 for the definition of ``domain``.
+
+Versions
+--------
+
+Though all NURLs are syntactically compatible some semantic differences are allowed.
+These differences are separated into distinct versions.
+
+Version 0
+---------
+
+A Foolscap fURL is considered the canonical definition of a version 0 NURL.
+Notably,
+the hash component is defined as the base32-encoded SHA1 hash of the DER form of an x509v3 certificate.
+A version 0 NURL is identified by the absence of the ``v=1`` fragment.
+
+Version 1
+---------
+
+The hash component of a version 1 NURL differs in three ways from the prior version.
+
+1. The hash function used is SHA3-224 instead of SHA1.
```

Comment:
I guess we should get our cryptographers to comment on this (or did we?)

Reply:
We didn't but I just asked.  Some *overall* points (not just about this hash selection) for consideration that were raised:

* Since the SPKI is purely public information, an attacker can construct a certificate containing the correct SPKI.  If the only thing another party does is verify the SPKI hash, they can be fooled into believing the attacker is the real server.  However, if the attacker presents this certificate during a TLS handshake then even though the SPKI hash will match the attacker will be unable to complete the handshake because they lack the corresponding private key.  I think I will make this property clear in the doc so it is explicit that the hash must be verified *and* a TLS handshake must be completed or there are no real security properties.
* "SHA3 is good but SHA2 is also still considered good".  Maybe "SHA3 has more protection against length-extension attacks".  Probably "most people also trust blake2, which is really fast".  (According to https://www.blake2.net/ probably about 3x faster).  I suspect hash speed isn't the primary concern for this use case but maybe it's not irrelevant either, esp if we want to consider mobile clients (according to a random slashdot article I found, SHA3 might be 5x slower than blake2 but take this with a grain of salt).  Note blake2 *just* some random other option that might be faster.  It was also a SHA3 candidate, but not the winner.

Reply:
Another random point in favor of Blake2 might be that it can produce a digest of any size from 1 to 64 bytes so we can very precisely select our usability/strength balance point.

Reply:
Another perspective: "Just pick sha256, life is too short to spend time on this."

Reply:
From that perspective, I'm inclined to stick with sha3-224 because there seems to be no really compelling argument to switch to anything else.